cpp/include/raft/core/detail/buffer_utils/owning_buffer_base.hpp

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <raft/core/device_type.hpp>
+#include <raft/core/execution_device_id.hpp>
+#include <raft/core/execution_stream.hpp>
+#include <type_traits>
+
+namespace raft {
+namespace detail {
+
+template <device_type D, typename T>
+struct owning_buffer {
+  owning_buffer() {}
+  owning_buffer(execution_device_id<D> device_id, std::size_t size, execution_stream stream) {}
+  auto* get() const { return static_cast<T*>(nullptr); }
```

Comment:
Should this be virtual instead? Will these `*_base` objects be directly used anywhere?

Reply:
No, the base class is not used anywhere directly. The corresponding classes for cpu and gpu just inherit and overload the base class functions.